services/venice-controller/src/main/java/com/linkedin/venice/controller/ProtocolVersionAutoDetectionService.java

```diff
@@ -0,0 +1,196 @@
+          }
+        }
+      }
+    }
+
+    /**
+     * Get the smallest local admin operation protocol version for all consumers in the given cluster.
+     * This will help to ensure that all consumers are on the same page regarding the protocol version.
+     *
+     * @param clusterName The name of the cluster to check.
+     * @return The smallest local admin operation protocol version for all consumers (parent + child controllers)
+     * in the cluster.
+     */
+    public long getSmallestLocalAdminOperationProtocolVersionForAllConsumers(String clusterName) {
+      // Map to store all consumers versions
+      Map<String, Map<String, Long>> regionToControllerToVersionMap = new ConcurrentHashMap<>();
+
+      // Get all versions for parent clusters
+      Map<String, Long> parentClusterVersions =
+          veniceParentHelixAdmin.getVeniceHelixAdmin().getAdminOperationVersionFromControllers(clusterName);
+      regionToControllerToVersionMap.put("parentRegion", parentClusterVersions);
+
+      // Get child controller clients for all regions
+      Map<String, ControllerClient> controllerClientMap =
+          veniceParentHelixAdmin.getVeniceHelixAdmin().getControllerClientMap(clusterName);
+
+      // Forward the request to all regions
+      for (Map.Entry<String, ControllerClient> entry: controllerClientMap.entrySet()) {
+        Map<String, Long> controllerUrlToVersionMap = getControllerUrlToVersionMap(clusterName, entry);
+        regionToControllerToVersionMap.put(entry.getKey(), controllerUrlToVersionMap);
+      }
+
+      LOGGER.info("All controller versions for cluster {}: {}", clusterName, regionToControllerToVersionMap);
+      return getSmallestVersion(regionToControllerToVersionMap);
+    }
+
+    private Map<String, Long> getControllerUrlToVersionMap(
+        String clusterName,
+        Map.Entry<String, ControllerClient> regionToControllerClient) {
+      ControllerClient controllerClient = regionToControllerClient.getValue();
+      AdminOperationProtocolVersionControllerResponse response =
+          controllerClient.getAdminOperationProtocolVersionFromControllers(clusterName);
+      if (response.isError()) {
+        throw new VeniceException(
+            "Failed to get admin operation protocol version from child controller " + regionToControllerClient.getKey()
+                + ": " + response.getError());
+      }
+      return response.getControllerUrlToVersionMap();
+    }
+
+    /**
+     * Get the smallest version from a map of versions.
+     *
+     * @param versions A map of versions, where the key is regionName,
+     *                and the value is another map of controllerUrl to admin operation protocol version.
+     * @return The smallest version as a long.
+     */
+    private long getSmallestVersion(Map<String, Map<String, Long>> versions) {
+      return versions.values().stream().flatMap(m -> m.values().stream()).min(Long::compare).orElse(Long.MAX_VALUE);
```

Comment:
Just in case, when comparing the smallest version number with upstream version number in ZK, can we short circuit and ignore MAX VALUE?
I am worried that case like below happens:
85 (from ZK) != MAX_VALUE, so we set version number in ZK to MAX_VALUE.

Reply:
Another nice catch. Updated the logic to update to filter out the MAX_VALUE. I wonder whether we can do more things in this case. The case it returns MAX_VALUE is very rare, because it always have the local admin version. 

I have failed the task if it cannot get the response from child or standby parents. Is there any case we want to move forward even when one of them failed to get version? One option I am considering: Failed to get version from parent standby, but get version well for child controllers.

Reply:
> Is there any case we want to move forward even when one of them failed to get version?
You are correct here. Let's keep the current fail-fast behavior.